iscsi/iscsi.go

```diff
@@ -397,45 +444,154 @@ func DisconnectVolume(c Connector) error {
 	return nil
 }
 
-// RemovePhysicalDevice removes device(s) sdx from a Linux host.
-func RemovePhysicalDevice(devices ...string) error {
-	debug.Printf("Removing scsi device %v.\n", devices)
+// getMountTargetDevice returns the device to be mounted among the configured devices
+func (c *Connector) getMountTargetDevice() (*Device, error) {
+	if len(c.Devices) > 1 {
+		// getMultipathDevice returns a multipath device for the configured targets if it exists
+		multipathDevice, err := getMultipathDevice(c.Devices)
+		if err != nil {
+			debug.Printf("mount target is not a multipath device: %v", err)
+			return nil, err
+		}
+		debug.Printf("mount target is a multipath device")
+		return multipathDevice, nil
+	}
+
+	if len(c.Devices) == 0 {
+		return nil, fmt.Errorf("could not find mount target device: connector does not contain any device")
+	}
+
+	return &c.Devices[0], nil
+}
+
+// IsMultipathEnabled check if multipath is enabled on devices handled by this connector
+func (c *Connector) IsMultipathEnabled() bool {
+	return c.MountTargetDevice.Type == "mpath"
+}
+
+// GetScsiDevices get SCSI devices from device paths
+// It will returns all SCSI devices if no paths are given
+func GetScsiDevices(devicePaths []string, strict bool) ([]Device, error) {
+	debug.Printf("Getting info about SCSI devices %s.\n", devicePaths)
+
+	deviceInfo, err := lsblk(devicePaths, strict)
+	if err != nil {
+		debug.Printf("An error occured while looking info about SCSI devices: %v", err)
+		return nil, err
+	}
+
+	return deviceInfo.BlockDevices, nil
+}
+
+// GetIscsiDevices get iSCSI devices from device paths
+// It will returns all iSCSI devices if no paths are given
+func GetIscsiDevices(devicePaths []string, strict bool) (devices []Device, err error) {
+	scsiDevices, err := GetScsiDevices(devicePaths, strict)
+	if err != nil {
+		return
+	}
+
+	for i := range scsiDevices {
+		device := &scsiDevices[i]
+		if device.Transport == "iscsi" {
+			devices = append(devices, *device)
+		}
+	}
+
+	return
+}
+
+// lsblk execute the lsblk commands
+func lsblk(devicePaths []string, strict bool) (*deviceInfo, error) {
+	flags := []string{"-J", "-o", "NAME,HCTL,TYPE,TRAN,SIZE"}
```

Comment:
When I tested "lsblk -J -o "NAME,HCTL,TYPE,TRAN,SIZE" /dev/sda" under RHEL 7.9, which uses an older version lsblk from util-linux 2.23.2, this command fails. The -J JSON option was added during a later release. Some of our deployments in Thailand use older RHEL versions, so I suggest that this function use lsblk -r (raw) option and the parsing change to handle the raw format if possible.